datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -737,19 +761,30 @@ private void uninitializeTasks(List<DatastreamTask> tasks) {
     tasks.forEach(_cpProvider::unassignDatastreamTask);
   }
 
-  private void initializeTask(DatastreamTask task) {
-    DatastreamTaskImpl taskImpl = (DatastreamTaskImpl) task;
-    assignSerdes(taskImpl);
+  private void initializeTask(DatastreamTask task, Set<DatastreamTask> failedDatastreamTasks, boolean retryAndSaveError) {
+    try {
+      DatastreamTaskImpl taskImpl = (DatastreamTaskImpl) task;
+      assignSerdes(taskImpl);
 
-    boolean customCheckpointing = getCustomCheckpointing(task);
-    TransportProviderAdmin tpAdmin = _transportProviderAdmins.get(task.getTransportProviderName());
-    TransportProvider transportProvider = tpAdmin.assignTransportProvider(task);
-    EventProducer producer =
-        new EventProducer(task, transportProvider, _cpProvider, _eventProducerConfig, customCheckpointing);
+      boolean customCheckpointing = getCustomCheckpointing(task);
+      TransportProviderAdmin tpAdmin = _transportProviderAdmins.get(task.getTransportProviderName());
+      TransportProvider transportProvider = tpAdmin.assignTransportProvider(task);
+      EventProducer producer = new EventProducer(task, transportProvider, _cpProvider, _eventProducerConfig, customCheckpointing);
 
-    taskImpl.setEventProducer(producer);
-    Map<Integer, String> checkpoints = producer.loadCheckpoints(task);
-    taskImpl.setCheckpoints(checkpoints);
+      taskImpl.setEventProducer(producer);
+      Map<Integer, String> checkpoints = producer.loadCheckpoints(task);
+      taskImpl.setCheckpoints(checkpoints);
+    } catch (Exception e) {
+      if (e instanceof InterruptedException) {
+        _log.warn("onAssignmentChange call got interrupted", e);
+        throw e;
+      }
+      _log.warn("Failed to initialize {} task", task.getDatastreamTaskName());
+      if (retryAndSaveError) {
+        _eventQueue.put(CoordinatorEvent.createHandleInstanceErrorEvent(ExceptionUtils.getRootCauseMessage(e)));
+        failedDatastreamTasks.add(task);
```

Comment:
could we try and mock this scenario in a test?

Reply:
I have added a unit test for this in TestCoordinator. What I am trying to do there in to fail TransportProvider creation once and which should result in the event to be queued again and the next time TransportProvider creation goes through and the task is successfully initialized.